### PR TITLE
Add a random delay of 2 - 7 seconds between downloads

### LINKED ISSUE #51: Feature: Random delays
This is needed to avoid IP banning

### PR DESCRIPTION
**Pull request by SpyrosRoum**
This closes #51 but the performance drops a lot. Maybe a lower base delay would be possible?

### CODE REVIEW

**Comment by Skallwar** [comment]
> Maybe a lower base delay would be possible?
I think the best option will be to add a command line argument for DELAY_BASE_MILLIS and another one for the random range. This way, on some website with no rate limit the performance will not drop

File: src/scraper.rs [hunk 1]
```diff
@@ -168,6 +175,11 @@ impl Scraper {
                             Ok((url, depth)) => {
                                 counter = 0;
                                 Scraper::handle_url(&self_clone, &tx, url, depth);
+                                // Generate a custom delay from 2, to 7 seconds
+                                // Base delay is already 2 seconds
+                                let delay_added_millis = rng.gen_range(0, 6) * 1000;
+                                let delay_duration = time::Duration::from_millis(DELAY_BASE_MILLIS + delay_added_millis);
```
Comment (Skallwar) [inline src/scraper.rs:181]: Maybe make this a sub function ?

**Comment by SpyrosRoum** [comment]
I never even considered making it a command line argument but you are right, it's a good idea!
Though I'm thinking it may be more user friendly to make the user put seconds instead of milliseconds. That way it would be clearer what is needed (milliseconds of delay would probably still get you banned) and the user has to type less.
What do you think?

**Comment by CohenArthur** [comment]
> What do you think?

I think it's a good idea, especially if the usual delay is between 2 and 7 seconds

**Comment by Skallwar** [comment]
> What do you think?

Yes good idea :+1:

File: src/scraper.rs (hunk 1 above)
Comment (SpyrosRoum) [reply src/scraper.rs]: That's done now

**Comment by SpyrosRoum** [comment]
Alright, I added two command line arguments `low` and `high`
I also added some default values for both and a check in case low > higher

**Comment by Skallwar** [comment]
> Alright, I added two command line arguments `low` and `high`

We could see `low` as a base delay and `high` as the upper bound of a random delay (0 to high). This imply that `--low` would be rename as `--delay` (no short as `-d` is use for the depth) and `--high` as `--random-range`. It makes more sense to me. 

> Ah I forgot to update the test. I guess it would be a good idea to also add tests for various amounts of low/high? Or 0 and 0 would be enough?

This kind of behavior is hard to unit test. Maybe add a functional test ? 2 tests, one for 0 0 and one for more than 0 0 should be enough

**Comment by SpyrosRoum** [comment]
I'm not sure if I understand `--random-range` correctly. Would you do something like `--delay 2 --random-range 10` and that would add a delay from 2 - 10 seconds?

**Comment by Skallwar** [comment]
> I'm not sure if I understand `--random-range` correctly. Would you do something like `--delay 2 --random-range 10` and that would add a delay from 2 - 10 seconds?

More like 2 + rand(0, 10)

File: src/args.rs [hunk 2]
```diff
@@ -29,6 +29,14 @@ pub struct Args {
     ///Show all logs
     #[structopt(short, long, help = "Enable more information regarding the scraping process")]
     pub verbose: bool,
+
+    /// The least seconds of delay between downloads
+    #[structopt(long, default_value = "0", help="Add a delay in seconds between downloads to reduce the likelihood of getting banned")]
+    pub delay: u64,
+
+    /// The max seconds of delay between downloads
+    #[structopt(long, default_value = "0", help="Generate an extra random delay between downloads up to this number. This is added to the base delay seconds")]
```
Comment (Skallwar) [inline src/args.rs:38]:
```suggestion
    #[structopt(long, default_value = "0", help="Generate an extra random delay between downloads, from 0 to this number. This is added to the base delay seconds")]
```

File: src/scraper.rs [hunk 3]
```diff
@@ -177,6 +182,19 @@ impl Scraper {
         .unwrap();
     }
 
+    /// Sleep the thread for a variable amount of seconds to avoid getting banned
+    fn sleep(&self, rng: &mut rand::rngs::ThreadRng) {
+        let base = self.args.delay;
+        let delay_range = self.args.random_range;
```
Comment (Skallwar) [inline src/scraper.rs:188]:
```suggestion
        let random_range = self.args.random_range;
```

File: src/scraper.rs [hunk 4]
```diff
@@ -177,6 +182,19 @@ impl Scraper {
         .unwrap();
     }
 
+    /// Sleep the thread for a variable amount of seconds to avoid getting banned
+    fn sleep(&self, rng: &mut rand::rngs::ThreadRng) {
+        let base = self.args.delay;
```
Comment (Skallwar) [inline src/scraper.rs:187]:
```suggestion
        let base_delay = self.args.delay;
```

File: src/scraper.rs [hunk 5]
```diff
@@ -177,6 +182,19 @@ impl Scraper {
...
+        let delay_range = self.args.random_range;
+
+        if base == 0 && delay_range == 0 { return; }
+
+        // delay_range+1 because gen_range is exclusive on the upper limit
+        let rand_delay_secs = rng.gen_range(0, delay_range+1);
+        let delay_duration = time::Duration::from_secs(base+rand_delay_secs);
```
Comment (CohenArthur) [inline src/scraper.rs:194]:
```suggestion
        let delay_duration = time::Duration::from_secs(base + rand_delay_secs);
```
Same as above

File: src/scraper.rs [hunk 6]
```diff
@@ -177,6 +182,19 @@ impl Scraper {
...
 
+    /// Sleep the thread for a variable amount of seconds to avoid getting banned
+    fn sleep(&self, rng: &mut rand::rngs::ThreadRng) {
+        let base = self.args.delay;
+        let delay_range = self.args.random_range;
+
+        if base == 0 && delay_range == 0 { return; }
+
+        // delay_range+1 because gen_range is exclusive on the upper limit
+        let rand_delay_secs = rng.gen_range(0, delay_range+1);
```
Comment (CohenArthur) [inline src/scraper.rs:193]:
```suggestion
        let rand_delay_secs = rng.gen_range(0, delay_range + 1);
```

Add more spaces between operators

**Comment by SpyrosRoum** [comment]
Oops, I shouldn't blindly accept things haha

**Comment by Skallwar** [comment]
> Oops, I shouldn't blindly accept things haha

You just need to change the variable names everywhere, no big deal

**Comment by Skallwar** [comment]
Nice work :rocket:

**Pull request merged**